### ISSUE TITLE
Camera not working 

### ISSUE THREAD

**Issue by RaghidJ** [processing/processing-video#168]
I'm not sure what is happening. The same code works sometimes and at others it doesn't: I can't connect to my camera. Is there any way to fix this issue? Or is it a known bug ? Thanks
`[
![Processing problem](https://user-images.githubusercontent.com/55320529/98692000-13d23600-2388-11eb-8ded-81e1ca114c08.png)
](url)`

 

## Steps to Reproduce

1.
2.
3.

## Your Environment

 

* Processing version:
* Operating System and OS version:
* Other information:

**Comment by companje** [comment]
I have the same issue ("The same code works sometimes and at others it doesn't"). Running processing-3.5.4 on a fresh Windows 10 Pro PC with the default Processing Video library imported. I tried different Camera's: Logitech C920 and Kurokesu c1 pro. Both cameras work fine in the Windows Camera application but have unreliable results in Processing environment. Sometimes it works, sometimes it doesn't.

![98692000-13d23600-2388-11eb-8ded-81e1ca114c08](https://user-images.githubusercontent.com/156066/102650052-8cd46280-416a-11eb-8b0e-5794f24d6cdd.png)
```
Could not find any devices
IllegalStateException: Could not find any devices
```

**Comment by lmtanco** [comment]
We have been noticing this behaviour consistently since the upgrade from 3.5.3 to 3.5.4 in Windows 10. We also tried with 4a3 today: same behaviour as 3.5.4.

The issue is easy to reproduce, with two tests.  

1. Use Capture with laptop's internal camera. println(Capture.list()) gives nothing (trying to instance a Capture object crashes gives "Could not find any devices".) 
2. Plug in external camera: println(Capture.list()) gives both cameras, the previously ignored internal camera and the new one. (trying to instance a Capture object with any of the cameras works fine.) 

I attach screenshots for both tests:  

Before plugging external cam: 

![image](https://user-images.githubusercontent.com/1093084/105690356-2c200e80-5efc-11eb-80df-cbb938f0def1.png)

After pluggin external cam: 

![image](https://user-images.githubusercontent.com/1093084/105690496-5eca0700-5efc-11eb-8f39-afb69e1903ea.png)

**Comment by lucasfredericks** [comment]
I have run into this problem a few different times. I think it may have to do with cameras which support both yuyv and mjpeg. Anyway, I finally found [this](https://discourse.processing.org/t/processing-cant-find-the-camera-video-libraries-dont-work/25128/13) solution. Instead of Capture.list(), use a custom gstreamer pipeline. It defaults to yuyv, which is annoying if you wanted hd video at more than 2 fps.  After much trial and error, here are some pipelines that worked for me. 
Generic yuyv: 
`cam = new Capture(this, "pipeline:autovideosrc");`
yuyv with some parameters:
`cam = new Capture(this, 640, 480, "pipeline: ksvideosrc device-index=0 ! video/x-raw,width=640,height=480");`
mjpeg: 
`cam = new Capture(this, 1920,1080, "pipeline: ksvideosrc device-index=0! image/jpeg, width=1920, height=1080, framerate=30/1 ! jpegdec ! videoconvert");`
Note that you should specify your resolution twice so that Processing and gstreamer are on the same page. It's not necessary in all cases, but you might save yourself a headache.

**Comment by ToCenTek** [comment]
> I have run into this problem a few different times. I think it may have to do with cameras which support both yuyv and mjpeg. Anyway, I finally found [this](https://discourse.processing.org/t/processing-cant-find-the-camera-video-libraries-dont-work/25128/13) solution. Instead of Capture.list(), use a custom gstreamer pipeline. It defaults to yuyv, which is annoying if you wanted hd video at more than 2 fps. After much trial and error, here are some pipelines that worked for me.
> Generic yuyv:
> `cam = new Capture(this, "pipeline:autovideosrc");`
> yuyv with some parameters:
> `cam = new Capture(this, 640, 480, "pipeline: ksvideosrc device-index=0 ! video/x-raw,width=640,height=480");`
> mjpeg:
> `cam = new Capture(this, 1920,1080, "pipeline: ksvideosrc device-index=0! image/jpeg, width=1920, height=1080, framerate=30/1 ! jpegdec ! videoconvert");`
> Note that you should specify your resolution twice so that Processing and gstreamer are on the same page. It's not necessary in all cases, but you might save yourself a headache.

Great! This solves my problem! Thank you so much!

**Comment by EinsMalte** [comment]
Yeah cool. Using Processing 4.0 beta and GStreamer 1.16.2 I get following errors:

`BaseSrc: [ksvideosrc0] : No supported formats found
BaseSrc: [ksvideosrc0] : not negotiated
BaseSrc: [ksvideosrc0] : Internal data stream error.`
